♻️✅ Extract `isFieldDefault` helper

xrav3nz opened the pull request:
Extract an common helper that can be used for form field dirtiness
check.

This is part of #22534.

/cc @cvializ @GoTcWang

xrav3nz commented on extensions/amp-form/0.1/test/test-utils.js:
```diff
@@ -0,0 +1,116 @@
+
+  describe('isFieldDefault', () => {
+    it("returns if a text field's value matches its default value", () => {
+      const html = '<input type="text" value="default">';
```
Elements are inserted as HTML so that the `defaultValue` property is generated correctly.

GoTcWang commented on extensions/amp-form/0.1/test/test-utils.js:
```diff
@@ -0,0 +1,116 @@
+      element.value = 'not default';
+      expect(isFieldDefault(element)).to.be.false;
+
+      element.value = 'default';
```
Ideally one unit test should only test one behavior, so this should be split into two tests, one for default and one for not.

In this way we can also name the test with more information like: returns true if the field is default value.

You can also add nested `describes` into one tests (and possibly `setUp`s, which I am not so sure), so you can bundle those two tests in one description ("text fields" for example) 

Same below.

GoTcWang commented on extensions/amp-form/0.1/utils.js:
```diff
@@ -0,0 +1,40 @@
+    case 'radio':
+      return field.checked === field.defaultChecked;
+    default:
+      return field.value === field.defaultValue;
```
Out the scope of this PR, what's the possible other elements?

Why not make it an exhaustive switch and throw some errors in default?

cvializ replied:
Agreed it does seem like something that could be made clearer, but it's better left for a future PR since we're only factoring out this method to be reused now.

GoTcWang replied:
+1 on doing that in a separate PR, thanks!

GoTcWang commented on extensions/amp-form/0.1/test/test-utils.js:
```diff
@@ -24,93 +24,134 @@ describes.realWin('amp-form-utils', {}, env => {
+      });
+
+      it("returns true if text field's value matches its default value", () => {
+        expect(isFieldDefault(textField)).to.be.true;
```
You can explicitly call 
textField.value = textField.default;
in such tests, so the tests can explain themselves.

And same below.

> the test's body should be complete in that it contains all of the relevant information that a user needs to understand a test without having to track down external references and definitions.

> A test must never explicitly rely on behavior defined in setUp, or else it becomes impossible to understand the test method by itself.

> A test should specify only the relevant parameters, but also all the relevant parameters, avoiding any implicit dependency on unrelated configuration.

GoTcWang commented on extensions/amp-form/0.1/test/test-utils.js:
```diff
@@ -53,6 +54,7 @@ describes.realWin('amp-form-utils', {}, env => {
       });
 
       it("returns true if textarea's value matches its default value", () => {
+        textarea.value = 'default';
```
It's okay to use a hardcoded string here as it's pretty much self-explained, but why not just use textarea.defaultValue? Would that work?

xrav3nz replied:
I deliberately chose to use a string value here, as I want the default value to not be empty. I could extract that into a named constant if that works better for you?

GoTcWang replied:
Okay I see, that should be fine.
These tests are pretty simple so I won't push further. But technically, these tests are still relying on the "hidden" fact that the default value is string 'default'. When you are working on some messier unit tests, you might consider moving those logic from `beforeEach` to tests themselves.

When writing unit tests, then most important thing is whether the test is clear to readers (i.e. if there are hidden facts) and then the cleanest/simplest of the tests.

cvializ commented on extensions/amp-form/0.1/utils.js:
```diff
@@ -0,0 +1,40 @@
+ * @param {!Element} field
+ * @return {boolean}
+ */
+export function isFieldDefault(field) {
```
nit: Let's put this method in `src/form.js` with the other form utility methods

xrav3nz replied:
For sure! Mind if I do that after #22536 is merged? Not sure if there could be merge conflicts.

cvializ commented on extensions/amp-form/0.1/test/test-utils.js:
```diff
@@ -0,0 +1,164 @@
+
+      beforeEach(() => {
+        const html = '<input type="text" value="default">';
+        doc.body.insertAdjacentHTML('afterbegin', html);
```
Typically we construct the elements with `doc.createElement` and `setAttribute` instead of relying on HTML-setting methods and properties like `innerHTML` and `insertAdjacentHTML`. Let's do the same for all these tests

xrav3nz replied:
Elements are inserted as HTML so that the `defaultValue` property is generated correctly. I tried the other approach at first, but `defaultValue` would always be empty. I think this is because:

> `defaultValue` returns the default value as originally **specified in the HTML** that created this object.

cvializ replied:
TIL, thanks! Let's add a comment linking/explaining that so future readers have all the information they need

xrav3nz replied:
Sorry, I was wrong here! 

I experimented with `element.value = ...` which does not work, but `element.setAttribute('value', ...)` actually sets both the DOM the `defaultValue` property correctly. Will do a small PR to not manipulate the HTML directly.

cvializ commented on extensions/amp-form/0.1/test/test-utils.js:
```diff
@@ -0,0 +1,164 @@
+        expect(isFieldDefault(textarea)).to.be.true;
+      });
+
+      it("returns false if textarea's value does not match its default value", () => {
```
nit: I think our linter usually catches this, but the convention for the amphtml code base is to use single quotes for

xrav3nz replied:
There is a single quote in the text `textarea's value`, so I used double quotes in this case. Is this okay?

cvializ replied:
Oh I see now, I guess if the linter allows it then it's fine. Thanks

xrav3nz commented:
Rebased off latest master, moved the helper to `src/form.js`, and added a comment on why the test elements are inserted as HTML (not sure if the comments are too verbose?).

cvializ commented:
I've seen code authors put a comment near the first instance of an unusual pattern, but these comments are good as they are. Thanks for making those changes
